relayer/chain/ethereum/listener.go

```diff
@@ -16,28 +18,37 @@ import (
 	"github.com/snowfork/polkadot-ethereum/relayer/chain"
 	"github.com/snowfork/polkadot-ethereum/relayer/chain/ethereum/syncer"
 	"github.com/snowfork/polkadot-ethereum/relayer/contracts/outbound"
+	"github.com/snowfork/polkadot-ethereum/relayer/contracts/polkadotrelaychainbridge"
+	"github.com/snowfork/polkadot-ethereum/relayer/contracts/validatorregistry"
+	"github.com/snowfork/polkadot-ethereum/relayer/store"
 )
 
 const MaxMessagesPerSend = 10
 
 // Listener streams the Ethereum blockchain for application events
 type Listener struct {
-	config    *Config
-	conn      *Connection
-	contracts []*outbound.Contract
-	messages  chan<- []chain.Message
-	headers   chan<- chain.Header
-	log       *logrus.Entry
+	config                   *Config
+	conn                     *Connection
+	db                       *store.Database
+	contracts                []*outbound.Contract
+	polkadotRelayChainBridge *polkadotrelaychainbridge.Contract
+	validatorRegistry        *validatorregistry.Contract
+	messages                 chan<- []chain.Message
+	beefyMessages            chan<- store.DatabaseCmd
+	headers                  chan<- chain.Header
+	log                      *logrus.Entry
 }
 
-func NewListener(config *Config, conn *Connection, messages chan<- []chain.Message, headers chan<- chain.Header, contracts []*outbound.Contract, log *logrus.Entry) (*Listener, error) {
+func NewListener(config *Config, conn *Connection, db *store.Database, messages chan<- []chain.Message, beefyMessages chan<- store.DatabaseCmd, headers chan<- chain.Header, contracts []*outbound.Contract, log *logrus.Entry) (*Listener, error) {
```

Comment:
need to think about the cold-start flow, ie, if relayer crashes in middle of something and is restarted, can it recover the state of its DB by reading from both chains? can think about/do in follow up PR/task though

Reply:
Yea called this one out in the PR description, it's something I've been thinking about as well. It should be possible to sync from an arbitrary previous block number and properly build up DB state. Can't go through the same sequence of tx sending/checks as we do in live-mode though, so will need some sort of db syncing mode.

Reply:
yeh - can u add an issue for it?

Reply:
Included this in issue https://github.com/Snowfork/polkadot-ethereum/issues/340